### DISCUSSION TITLE
Helper Function to Export Predicted Links to Graph Format

### DISCUSSION THREAD

**Discussion by ozlander** [pyg-team/pytorch_geometric#2246] [Q&A] [answered]
Thank you for making the fantastic Pytorch Geometric! It's already helped me answer real questions. A key difficulty when coming from far less technical fields is the need to explore individual cases (e.g., the social sciences). For example, I'm exploring examples/tgn.py and trying to export the links predicted by the model (y_pred) to an interpretable format. Is there an efficient way to do this? I.e., translate edge_index back to the original input labels for nodes and edges and save the constructed y_pred and y_true graphs into say a NetworkX understandable format.

I realise this is not a task that is necessarily of interest to the usual highly technical field utilising PG, but I do believe a simple example of how to do this particular task will help to bridge applications of the models into non-technical fields that are interested in using the technology in qualitative ways. Thank you very much for your time and any assistance you can provide in achieving this goal.

**Comment by rusty1s** [comment, marked as answer]
Thanks for your interest. You are absolutely right. In general, we expect the user to handle the mapping from "semantic nodes" to node indices and vice versa. Since the user creates the data, this mapping should be known to him/her.

In order to achieve this, one may hold a dictionary which maps semantics to node indices:
```python
mapping = { 'node1': 0, 'node2': 1, 'node3': 2, ... }
```
which can be used to query a model (in this case for node classification):
```
out = model(x, edge_index)
# Get the prediction for 'node2':
out[mapping['node2']]
``` 

I have no problem with integrating an example to the library to show-case this.

**Reply by ozlander** [reply]
Hi @rusty1s, thank you for the quick response. I think I follow the logic behind the mapping, which I will experiment with now. To make sure I understand the rest of the implementation, is it correct to interpret the output of `pos_out = link_pred(z[assoc[src]], z[assoc[pos_dst]])` to mean the probabilities of edges as predicted by the model?

In other words, if `edge_index` gives me the following: `tensor([[418, 402, 391, ..., 7, 6, 154], [0, 0, 0, ..., 429, 429, 429]])`, and `pos_out` gives me `tensor([[1.5060],[-1.9640],[-0.3152], ..., [-1.5560], [-2.0355], [0.9326]])`, I can take this to mean that from the test data, the indexed directional edges 418->0, 402->0, and 154->429 have predicted probabilities of 1.5060, -1.9640, and 0.9326, respectively. So in truth, if these are 1, 0, 1, the model would be 100% correct in its predictions (just on these three edges as an example)?

Please forgive the over-simplification and extremely dumb question - learning is definitely a process! Assuming this is indeed correct, and if `edge_index` is mapped to the original data as you suggest, I should be able to create any traditional network format of the predicted edges with that information put together.

**Reply by rusty1s** [reply]
Yes, exactly. In this case, you may want to apply a `sigmoid` function to the output to convert the values to probabilities in range `[0, 1]`.

**Reply by ozlander** [reply]
Thank you again. I think I've got that part figured out now. tgn.py seems to be a little bit more complex than standard models. I see `n_id, edge_index, e_id = neighbor_loader(n_id)` produces an edge index in batches during the test phase, which is where I want to convert predictions back to original labels. I'm able to get the original labels through some terrible code I'm using earlier on to refer back to and hopefully construct a mapping:

```
edges = []
i = 0
for src in test_data.src:
    dst = test_data.dst[i]
    edges.append([src.item(), dst.item()])
    i += 1
```

How I can map `edge_index` back to this rudimentary list seems a bit more difficult than anticipated. Am I missing something simple in getting from `edge_index` to the original labels of the edges?

**Reply by rusty1s** [reply]
You can map the subgraph `edge_index` back to original indices via `n_id`:
```python
global_edge_index = n_id[edge_index]
```
which lets you convert it into a list:
```python
edges = global_edge_index.t().tolist()
```
